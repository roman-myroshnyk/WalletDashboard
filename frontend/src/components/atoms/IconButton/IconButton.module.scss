@import '@/styles/colors.scss';

.iconButton {
  width: 36px;
  height: 36px;
  margin: 0;
  padding: 0;

  color: $gray;

  background-color: transparent;
  border: 0;
  border-radius: 8px;
}

.iconButton:disabled {
  color: $darkGray;
}
