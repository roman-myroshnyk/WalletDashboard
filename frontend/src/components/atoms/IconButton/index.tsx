import { FC, ReactNode } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import * as colors from '@/styles/colors';
import styles from './IconButton.module.scss';

export interface IProps {
  hasInitialAnimation?: boolean;
  isShown: boolean;
  disabled: boolean;
  label: string;
  onClick: () => void;
  children:ReactNode;
}

const variants = {
  whileHover: {
    color: colors.BLUE,
    scale: 1.1,
    transition: {
      duration: 0.4,
    },
  },
  whileHoverDisabled: {
    scale: 1,
  },
  whileTap: {
    backgroundColor: colors.LIGHT_BLACK,
    transition: {
      duration: 0.1,
    },
  },
  whileTapDisabled: {
    backgroundColor: 'transparent',
  },
};

const IconButton:FC<IProps> = ({
  hasInitialAnimation, isShown, disabled, label, onClick, children,
}) => (
  <div>
    <AnimatePresence initial={hasInitialAnimation} exitBeforeEnter>
      {isShown && (
      <motion.button
        area-label={label}
        onClick={onClick}
        disabled={disabled}
        className={styles.iconButton}
        whileHover={disabled ? 'whileHoverDisabled' : 'whileHover'}
        whileTap={disabled ? 'whileTapDisabled' : 'whileTap'}
        variants={variants}
      >
        {children}
      </motion.button>
      )}
    </AnimatePresence>
  </div>
);

IconButton.defaultProps = {
  hasInitialAnimation: false,
};

export default IconButton;
