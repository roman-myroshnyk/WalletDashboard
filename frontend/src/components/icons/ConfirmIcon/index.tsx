import { FC } from 'react';
import { motion } from 'framer-motion';
import { getToggleAnimations } from '@/animations/ToggleAnimations';

export interface IProps {
  height?: number;
  width?: number;
  enterDelay?: number;
  exitDelay?: number;
 }

const ConfirmIcon:FC<IProps> = ({
  height, width, enterDelay, exitDelay,
}) => {
  const variants = getToggleAnimations(enterDelay, exitDelay);
  return (
    <motion.svg
      xmlns="http://www.w3.org/2000/svg"
      className="h-6 w-6"
      fill="none"
      width={width}
      height={height}
      viewBox="0 0 24 24"
      stroke="currentColor"
      strokeWidth={2}
      initial="exit"
      animate="enter"
      exit="exit"
      variants={variants.svg}
    >
      <motion.path
        variants={variants.path}
        strokeLinecap="round"
        strokeLinejoin="round"
        d="M5 13l4 4L19 7"
      />

    </motion.svg>
  );
};

ConfirmIcon.defaultProps = {
  height: 24,
  width: 24,
  enterDelay: 0,
  exitDelay: 0,
};

export default ConfirmIcon;
