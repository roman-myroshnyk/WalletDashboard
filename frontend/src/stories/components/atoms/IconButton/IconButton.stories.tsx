import { Meta, Story } from '@storybook/react/types-6-0';
import IconButton, { IProps as IIconButtonProps } from '@/atoms/IconButton';
import CancelIcon from '@/icons/CancelIcon';
import ConfirmIcon from '@/icons/ConfirmIcon';
import EditIcon from '@/icons/EditIcon';

export default {
  title: 'atoms/IconButton',
  component: IconButton,
  argTypes: {
    isShown: {
      contorols: 'boolean',
    },
    disabled: {
      controls: 'boolean',
    },
    onClick: {
      action: 'onClick: icon button clicked',
    },
  },
} as Meta;

const CancelIconButtonTemplate:Story<IIconButtonProps> = (args) => (
  <IconButton {...args}>
    <CancelIcon />
  </IconButton>
);

export const CancelIconButton = CancelIconButtonTemplate.bind({});

CancelIconButton.args = {
  isShown: true,
  disabled: false,
};

const ConfirmIconButtonTemplate:Story<IIconButtonProps> = (args) => (
  <IconButton {...args}>
    <ConfirmIcon />
  </IconButton>
);

export const ConfirmIconButton = ConfirmIconButtonTemplate.bind({});

ConfirmIconButton.args = {
  isShown: true,
  disabled: false,
};

const EditIconButtonTemplate:Story<IIconButtonProps> = (args) => (
  <IconButton {...args}>
    <EditIcon />
  </IconButton>
);

export const EditIconButton = EditIconButtonTemplate.bind({});

EditIconButton.args = {
  isShown: true,
  disabled: false,
};
