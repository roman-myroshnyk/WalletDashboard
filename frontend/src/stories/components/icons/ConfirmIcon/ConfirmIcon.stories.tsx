import { Meta, Story } from '@storybook/react/types-6-0';
import ConfirmIcon, { IProps } from '@/icons/ConfirmIcon';

export default {
  title: 'icons/ConfirmIcon',
  component: ConfirmIcon,
  argType: {
    height: {
      controls: 'number',
    },
    width: {
      controls: 'number',
    },
  },
} as Meta;

const Template:Story<IProps> = (args) => (
  <ConfirmIcon {...args} />
);

export const Default = Template.bind({});

Default.args = {
  height: 24,
  width: 24,
};
