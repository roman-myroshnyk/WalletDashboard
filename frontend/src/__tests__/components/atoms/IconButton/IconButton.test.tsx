import '../../../mocks/framer-motion';
import { fireEvent, render } from '@testing-library/react';
import IconButton from '@/atoms/IconButton';
import ConfirmIcon from '@/icons/ConfirmIcon';

describe('atoms/IconButton', () => {
  it('renders IconButton in normal state', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders IconButton in disabled state', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        hasInitialAnimation
        isShown={false}
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders IconButton isShown=true', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders IconButton isShown=false', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders IconButton hasInitialAnimation=false', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        hasInitialAnimation={false}
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('renders IconButton hasInitialAnimation=true', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        hasInitialAnimation
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    expect(container).toMatchSnapshot();
  });

  it('IconButton disabled=false onClick should fire', () => {
    const handleClick = jest.fn();
    const { getByRole } = render(
      <IconButton
        hasInitialAnimation={false}
        isShown
        disabled={false}
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    const button = getByRole('button');
    fireEvent.click(button);
    expect(handleClick).toBeCalledTimes(1);
  });

  it('IconButton disabled=true  onClick should not fire', () => {
    const handleClick = jest.fn();
    const { container } = render(
      <IconButton
        hasInitialAnimation
        isShown={false}
        disabled
        label="Icon button"
        onClick={handleClick}
      >
        <ConfirmIcon />
      </IconButton>,
    );
    fireEvent.click(container);
    expect(handleClick).toBeCalledTimes(0);
  });
});
