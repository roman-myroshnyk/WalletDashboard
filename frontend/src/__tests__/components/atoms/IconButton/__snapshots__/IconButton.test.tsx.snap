// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atoms/IconButton renders IconButton hasInitialAnimation=false 1`] = `
<div>
  <div>
    <button
      area-label="Icon button"
      class="iconButton"
      variants="[object Object]"
    >
      <svg
        animate="enter"
        aria-label="confirmIcon"
        class="h-6 w-6"
        exit="exit"
        fill="none"
        height="24"
        initial="exit"
        stroke="currentColor"
        stroke-width="2"
        variants="[object Object]"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          variants="[object Object]"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`atoms/IconButton renders IconButton hasInitialAnimation=true 1`] = `
<div>
  <div>
    <button
      area-label="Icon button"
      class="iconButton"
      variants="[object Object]"
    >
      <svg
        animate="enter"
        aria-label="confirmIcon"
        class="h-6 w-6"
        exit="exit"
        fill="none"
        height="24"
        initial="exit"
        stroke="currentColor"
        stroke-width="2"
        variants="[object Object]"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          variants="[object Object]"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`atoms/IconButton renders IconButton in disabled state 1`] = `
<div>
  <div />
</div>
`;

exports[`atoms/IconButton renders IconButton in normal state 1`] = `
<div>
  <div>
    <button
      area-label="Icon button"
      class="iconButton"
      variants="[object Object]"
    >
      <svg
        animate="enter"
        aria-label="confirmIcon"
        class="h-6 w-6"
        exit="exit"
        fill="none"
        height="24"
        initial="exit"
        stroke="currentColor"
        stroke-width="2"
        variants="[object Object]"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          variants="[object Object]"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`atoms/IconButton renders IconButton isShown=false 1`] = `
<div>
  <div>
    <button
      area-label="Icon button"
      class="iconButton"
      variants="[object Object]"
    >
      <svg
        animate="enter"
        aria-label="confirmIcon"
        class="h-6 w-6"
        exit="exit"
        fill="none"
        height="24"
        initial="exit"
        stroke="currentColor"
        stroke-width="2"
        variants="[object Object]"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          variants="[object Object]"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`atoms/IconButton renders IconButton isShown=true 1`] = `
<div>
  <div>
    <button
      area-label="Icon button"
      class="iconButton"
      variants="[object Object]"
    >
      <svg
        animate="enter"
        aria-label="confirmIcon"
        class="h-6 w-6"
        exit="exit"
        fill="none"
        height="24"
        initial="exit"
        stroke="currentColor"
        stroke-width="2"
        variants="[object Object]"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 13l4 4L19 7"
          stroke-linecap="round"
          stroke-linejoin="round"
          variants="[object Object]"
        />
      </svg>
    </button>
  </div>
</div>
`;
